--- a/uapi/linux/netlink.h_org	2020-03-13 20:10:50.000000000 +0100
+++ b/uapi/linux/netlink.h	2020-05-17 19:47:11.605853276 +0200
@@ -3,7 +3,7 @@
 #define __LINUX_NETLINK_H
 
 #include <linux/kernel.h>
-#include <linux/socket.h> /* for __kernel_sa_family_t */
+#include <linux/socket.h> /* for sa_family_t */
 #include <linux/types.h>
 
 #define NETLINK_ROUTE		0	/* Routing/device hook				*/
@@ -35,7 +35,7 @@
 #define MAX_LINKS 32		
 
 struct sockaddr_nl {
-	__kernel_sa_family_t	nl_family;	/* AF_NETLINK	*/
+	sa_family_t	nl_family;	/* AF_NETLINK	*/
 	unsigned short	nl_pad;		/* zero		*/
 	__u32		nl_pid;		/* port ID	*/
        	__u32		nl_groups;	/* multicast groups mask */
--- a/uapi/linux/rtnetlink.h_org	2020-03-13 20:10:50.000000000 +0100
+++ b/uapi/linux/rtnetlink.h	2020-05-17 19:46:59.149860334 +0200
@@ -414,7 +414,7 @@
 
 /* RTA_VIA */
 struct rtvia {
-	__kernel_sa_family_t	rtvia_family;
+	sa_family_t		rtvia_family;
 	__u8			rtvia_addr[0];
 };
 
