diff --git a/lib/expat.h b/lib/expat.h
index bb9cdedb..0bea3743 100644
--- a/lib/expat.h
+++ b/lib/expat.h
@@ -133,7 +133,7 @@ enum XML_Error {
   /* Added in 2.4.0. */
   XML_ERROR_AMPLIFICATION_LIMIT_BREACH,
   /* Added in 2.6.4. */
-  XML_ERROR_NOT_STARTED,
+  XML_ERROR_NOT_STARTED
 };
 
 enum XML_Content_Type {
@@ -1036,7 +1036,7 @@ enum XML_FeatureEnum {
   XML_FEATURE_GE,
   /* Added in Expat 2.7.2. */
   XML_FEATURE_ALLOC_TRACKER_MAXIMUM_AMPLIFICATION_DEFAULT,
-  XML_FEATURE_ALLOC_TRACKER_ACTIVATION_THRESHOLD_DEFAULT,
+  XML_FEATURE_ALLOC_TRACKER_ACTIVATION_THRESHOLD_DEFAULT
   /* Additional features must be added to the end of this enum. */
 };
 
