--- a/jpeglib.h
+++ b/jpeglib.h
@@ -951,6 +951,8 @@
 
 /* Standard data source and destination managers: stdio streams. */
 /* Caller is responsible for opening the file before and closing after. */
+/* fix unknown type name 'FILE' */
+#include <stdio.h>
 EXTERN(void) jpeg_stdio_dest JPP((j_compress_ptr cinfo, FILE * outfile));
 EXTERN(void) jpeg_stdio_src JPP((j_decompress_ptr cinfo, FILE * infile));
 

