diff --git a/Makefile.am b/Makefile.am
index 213a1fe..5934c60 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -118,7 +118,7 @@ libkmod_libkmod_internal_la_SOURCES = $(libkmod_libkmod_la_SOURCES)
 libkmod_libkmod_internal_la_LDFLAGS = $(AM_LDFLAGS) \
 	-Wl,--version-script=$(top_srcdir)/libkmod/libkmod.sym
 libkmod_libkmod_internal_la_DEPENDENCIES  = $(libkmod_libkmod_la_DEPENDENCIES)
-libkmod_libkmod_internal_la_LIBADD = $(libkmod_libkmod_la_LIBADD)
+libkmod_libkmod_internal_la_LIBADD = $(libkmod_libkmod_la_LIBADD) -lrt
 
 pkgconfig_DATA = libkmod/libkmod.pc
 noarch_pkgconfig_DATA = tools/kmod.pc
diff --git a/libkmod/libkmod.c b/libkmod/libkmod.c
index d1e5a82..19434b3 100644
--- a/libkmod/libkmod.c
+++ b/libkmod/libkmod.c
@@ -175,7 +175,7 @@ static char *get_kernel_release(const char *dirname)
 	if (uname(&u) < 0)
 		return NULL;
 
-	if (asprintf(&p, "%s/%s", dirname_default_prefix, u.release) < 0)
+	if (asprintf(&p, "%s", dirname_default_prefix) < 0)
 		return NULL;
 
 	return p;
diff --git a/tools/depmod.c b/tools/depmod.c
index 91bd191..a14e5d2 100644
--- a/tools/depmod.c
+++ b/tools/depmod.c
@@ -2982,17 +2982,14 @@ static int do_depmod(int argc, char *argv[])
 		cfg.kversion = un.release;
 	}
 
-	cfg.dirnamelen = snprintf(cfg.dirname, PATH_MAX, "%s%s/%s", root ?: "",
-				  module_directory, cfg.kversion);
+	cfg.dirnamelen = snprintf(cfg.dirname, PATH_MAX, "%s" MODULE_DIRECTORY "/", root ?: "");
 	if (cfg.dirnamelen >= PATH_MAX) {
 		ERR("Bad directory %s" MODULE_DIRECTORY "/%s: path too long\n",
 		    root ?: "", cfg.kversion);
 		goto cmdline_failed;
 	}
 
-	cfg.outdirnamelen = snprintf(cfg.outdirname, PATH_MAX, "%s%s/%s",
-				     out_root ?: (root ?: ""), module_directory,
-				     cfg.kversion);
+	cfg.outdirnamelen = snprintf(cfg.outdirname, PATH_MAX, "%s" MODULE_DIRECTORY "/", out_root ?: (root ?: ""));
 	if (cfg.outdirnamelen >= PATH_MAX) {
 		ERR("Bad directory %s" MODULE_DIRECTORY "/%s: path too long\n",
 		    out_root ?: (root ?: ""), cfg.kversion);
diff --git a/tools/modinfo.c b/tools/modinfo.c
index f1323c5..876d6cd 100644
--- a/tools/modinfo.c
+++ b/tools/modinfo.c
@@ -443,8 +443,7 @@ static int do_modinfo(int argc, char *argv[])
 			kversion = u.release;
 		}
 
-		n = snprintf(dirname_buf, sizeof(dirname_buf),
-			     "%s" MODULE_DIRECTORY "/%s", root, kversion);
+		n = snprintf(dirname_buf, sizeof(dirname_buf), "%s" MODULE_DIRECTORY "/", root);
 		if (n >= (int)sizeof(dirname_buf)) {
 			ERR("bad directory %s" MODULE_DIRECTORY "/%s: path too long\n",
 			    root, kversion);
diff --git a/tools/modprobe.c b/tools/modprobe.c
index 26b252c..b11aed5 100644
--- a/tools/modprobe.c
+++ b/tools/modprobe.c
@@ -919,8 +919,7 @@ static int do_modprobe(int argc, char **orig_argv)
 			}
 			kversion = u.release;
 		}
-		n = snprintf(dirname_buf, sizeof(dirname_buf),
-			     "%s" MODULE_DIRECTORY "/%s", root, kversion);
+		n = snprintf(dirname_buf, sizeof(dirname_buf), "%s" MODULE_DIRECTORY "/", root);
 		if (n >= (int)sizeof(dirname_buf)) {
 			ERR("bad directory %s" MODULE_DIRECTORY "/%s: path too long\n",
 			    root, kversion);
diff --git a/tools/static-nodes.c b/tools/static-nodes.c
index cb8afd3..f805af8 100644
--- a/tools/static-nodes.c
+++ b/tools/static-nodes.c
@@ -199,29 +199,19 @@ static int do_static_nodes(int argc, char *argv[])
 		goto finish;
 	}
 
-	r = snprintf(modules, sizeof(modules), MODULE_DIRECTORY "/%s/modules.devname",
-		     kernel.release);
+	r = snprintf(modules, sizeof(modules), MODULE_DIRECTORY "/modules.devname");
 	if (r >= (int)sizeof(modules)) {
-		fprintf(stderr,
-			"Error: could not open " MODULE_DIRECTORY
-			"/%s/modules.devname - path too long\n",
-			kernel.release);
+		fprintf(stderr, "Error: could not open " MODULE_DIRECTORY "/modules.devname - path too long\n");
 		ret = EXIT_FAILURE;
 		goto finish;
 	}
 	in = fopen(modules, "re");
 	if (in == NULL) {
 		if (errno == ENOENT) {
-			fprintf(stderr,
-				"Warning: " MODULE_DIRECTORY
-				"/%s/modules.devname not found - ignoring\n",
-				kernel.release);
+			fprintf(stderr, "Warning: " MODULE_DIRECTORY "/modules.devname not found - ignoring\n");
 			ret = EXIT_SUCCESS;
 		} else {
-			fprintf(stderr,
-				"Error: could not open " MODULE_DIRECTORY
-				"/%s/modules.devname - %m\n",
-				kernel.release);
+			fprintf(stderr, "Error: could not open " MODULE_DIRECTORY "/modules.devname\n");
 			ret = EXIT_FAILURE;
 		}
 		goto finish;
