diff --git a/Makefile.am b/Makefile.am
index d37b56d..0d576ce 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -107,7 +107,7 @@
 libkmod_libkmod_internal_la_LDFLAGS = $(AM_LDFLAGS) \
 	-Wl,--version-script=$(top_srcdir)/libkmod/libkmod.sym
 libkmod_libkmod_internal_la_DEPENDENCIES  = $(libkmod_libkmod_la_DEPENDENCIES)
-libkmod_libkmod_internal_la_LIBADD = $(libkmod_libkmod_la_LIBADD)
+libkmod_libkmod_internal_la_LIBADD = $(libkmod_libkmod_la_LIBADD) -lrt
 
 pkgconfig_DATA = libkmod/libkmod.pc
 noarch_pkgconfig_DATA = tools/kmod.pc
diff --git a/libkmod/libkmod.c b/libkmod/libkmod.c
index 213b424..56114a2 100644
--- a/libkmod/libkmod.c
+++ b/libkmod/libkmod.c
@@ -209,7 +209,7 @@ static char *get_kernel_release(const char *dirname)
 	if (uname(&u) < 0)
 		return NULL;
 
-	if (asprintf(&p, "%s/%s", dirname_default_prefix, u.release) < 0)
+	if (asprintf(&p, "%s", dirname_default_prefix) < 0)
 		return NULL;
 
 	return p;
diff --git a/tools/depmod.c b/tools/depmod.c
index 43fc354..a03ff8c 100644
--- a/tools/depmod.c
+++ b/tools/depmod.c
@@ -3010,13 +3010,9 @@ static int do_depmod(int argc, char *argv[])
 		cfg.kversion = un.release;
 	}
 
-	cfg.dirnamelen = snprintf(cfg.dirname, PATH_MAX,
-				  "%s" MODULE_DIRECTORY "/%s",
-				  root ?: "", cfg.kversion);
+	cfg.dirnamelen = snprintf(cfg.dirname, PATH_MAX, "%s" MODULE_DIRECTORY "/", root ?: "");
 
-	cfg.outdirnamelen = snprintf(cfg.outdirname, PATH_MAX,
-				     "%s" MODULE_DIRECTORY "/%s",
-				     out_root ?: (root ?: ""), cfg.kversion);
+	cfg.outdirnamelen = snprintf(cfg.outdirname, PATH_MAX, "%s" MODULE_DIRECTORY "/", out_root ?: (root ?: ""));
 
 	if (optind == argc)
 		all = 1;
diff --git a/tools/modinfo.c b/tools/modinfo.c
index cacc32d..4a8b003 100644
--- a/tools/modinfo.c
+++ b/tools/modinfo.c
@@ -450,8 +450,7 @@ static int do_modinfo(int argc, char *argv[])
 			}
 			kversion = u.release;
 		}
-		snprintf(dirname_buf, sizeof(dirname_buf), "%s" MODULE_DIRECTORY "/%s",
-			 root, kversion);
+		snprintf(dirname_buf, sizeof(dirname_buf), "%s" MODULE_DIRECTORY "/", root);
 		dirname = dirname_buf;
 	}
 
diff --git a/tools/modprobe.c b/tools/modprobe.c
index 5306bef..76686ef 100644
--- a/tools/modprobe.c
+++ b/tools/modprobe.c
@@ -987,9 +987,7 @@ static int do_modprobe(int argc, char **orig_argv)
 			}
 			kversion = u.release;
 		}
-		snprintf(dirname_buf, sizeof(dirname_buf),
-				"%s" MODULE_DIRECTORY "/%s", root,
-				kversion);
+		snprintf(dirname_buf, sizeof(dirname_buf), "%s" MODULE_DIRECTORY "/", root);
 		dirname = dirname_buf;
 	}
 
diff --git a/tools/static-nodes.c b/tools/static-nodes.c
index 5ef3743..73fba47 100644
--- a/tools/static-nodes.c
+++ b/tools/static-nodes.c
@@ -200,16 +200,14 @@ static int do_static_nodes(int argc, char *argv[])
 		goto finish;
 	}
 
-	snprintf(modules, sizeof(modules), MODULE_DIRECTORY "/%s/modules.devname", kernel.release);
+	snprintf(modules, sizeof(modules), MODULE_DIRECTORY "/modules.devname");
 	in = fopen(modules, "re");
 	if (in == NULL) {
 		if (errno == ENOENT) {
-			fprintf(stderr, "Warning: " MODULE_DIRECTORY "/%s/modules.devname not found - ignoring\n",
-				kernel.release);
+			fprintf(stderr, "Warning: " MODULE_DIRECTORY "/modules.devname not found - ignoring\n");
 			ret = EXIT_SUCCESS;
 		} else {
-			fprintf(stderr, "Error: could not open " MODULE_DIRECTORY "/%s/modules.devname - %m\n",
-				kernel.release);
+			fprintf(stderr, "Error: could not open " MODULE_DIRECTORY "/modules.devname - %m\n");
 			ret = EXIT_FAILURE;
 		}
 		goto finish;
