--- a/widget_image.c_org	2018-09-03 14:11:24.408545986 +0200
+++ b/widget_image.c	2018-09-03 14:30:36.745669510 +0200
@@ -116,13 +116,44 @@
 			error("Warning: Image %s: fopen(%s) failed: %s", Name, file, strerror(errno));
 			return;
 		}
-
 		Image->gdImage = gdImageCreateFromPng(fd);
 		fclose(fd);
 
-		if (Image->gdImage == NULL)
-		{
-			error("Warning: Image %s: CreateFromPng(%s) failed!", Name, file);
+		if (Image->gdImage == NULL) {
+			fd = fopen(file, "rb");
+			if (fd == NULL)
+			{
+				error("Warning: Image %s: fopen(%s) failed: %s", Name, file, strerror(errno));
+				return;
+			}
+			Image->gdImage = gdImageCreateFromJpeg(fd);
+			fclose(fd);
+		}
+
+		if (Image->gdImage == NULL) {
+			fd = fopen(file, "rb");
+			if (fd == NULL)
+			{
+				error("Warning: Image %s: fopen(%s) failed: %s", Name, file, strerror(errno));
+				return;
+			}
+			Image->gdImage = gdImageCreateFromGif(fd);
+			fclose(fd);
+		}
+
+		if (Image->gdImage == NULL) {
+			fd = fopen(file, "rb");
+			if (fd == NULL)
+			{
+				error("Warning: Image %s: fopen(%s) failed: %s", Name, file, strerror(errno));
+				return;
+			}
+			Image->gdImage = gdImageCreateFromBmp(fd);
+			fclose(fd);
+		}
+
+		if (Image->gdImage == NULL) {
+			error("Warning: Image %s: CreateFromPng/Jpeg/Gif/Bmp (%s) failed!", Name, file);
 			return;
 		}
 
