diff --git a/widget_ttf.c b/widget_ttf.c
index f971eca..18b7eef 100644
--- a/widget_ttf.c
+++ b/widget_ttf.c
@@ -133,15 +133,15 @@ static void widget_ttf_render(const char *Name, WIDGET_TTF * Image)
 
 		switch (toupper(mheight[0]))
 		{
-			case 'O': // old behavior
-				mtext = text;
+			case 'F': // fullsized for all possible chars
+				mtext = "[Äp}§|";
 				break;
-			case 'N': // ignore oversized chars, use for numbers etc
+			case 'N': // ignore oversized chars, use for umlauts etc.
 				mtext = "Äp";
 				break;
-			case 'F': // fullsized for all poss. chars (default)
+			case 'O': // autosize chars (default)
 			default:
-				mtext = "[Äp}§|";
+				mtext = text;
 				break;
 		}
 
