--- a/configure.ac
+++ b/configure.ac
@@ -265,7 +265,7 @@
 
 CURL_CHECK_COMPILER
 CURL_SET_COMPILER_BASIC_OPTS
-CURL_SET_COMPILER_DEBUG_OPTS
+dnl CURL_SET_COMPILER_DEBUG_OPTS
 CURL_SET_COMPILER_OPTIMIZE_OPTS
 CURL_SET_COMPILER_WARNING_OPTS
 
@@ -3149,6 +3149,7 @@
 
 AC_CHECK_SIZEOF(size_t)
 AC_CHECK_SIZEOF(long)
+AC_CHECK_SIZEOF(long long)
 AC_CHECK_SIZEOF(int)
 AC_CHECK_SIZEOF(short)
 CURL_CONFIGURE_LONG
--- a/Makefile.am
+++ b/Makefile.am
@@ -155,7 +155,7 @@
 bin_SCRIPTS = curl-config
 
 SUBDIRS = lib docs src include
-DIST_SUBDIRS = $(SUBDIRS) tests packages scripts
+DIST_SUBDIRS = $(SUBDIRS) packages
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
--- a/Makefile.in
+++ b/Makefile.in
@@ -623,7 +623,7 @@
 
 bin_SCRIPTS = curl-config
 SUBDIRS = lib docs src include
-DIST_SUBDIRS = $(SUBDIRS) tests packages scripts
+DIST_SUBDIRS = $(SUBDIRS) packages
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
 LIB_VAUTH_CFILES = vauth/vauth.c vauth/cleartext.c vauth/cram.c         \
