--- a/gio/Makefile.am
+++ b/gio/Makefile.am
@@ -246,7 +246,7 @@
 platform_deps += win32/libgiowin32.la
 endif
 
-SUBDIRS += . tests
+SUBDIRS += .
 
 if HAVE_FAM
 SUBDIRS += fam
--- a/glib/Makefile.am
+++ b/glib/Makefile.am
@@ -33,8 +33,8 @@
 MAYBE_PCRE = pcre 
 endif
 
-SUBDIRS = libcharset $(PRINTF_SUBDIR) $(MAYBE_PCRE) update-pcre . tests
-DIST_SUBDIRS = libcharset gnulib pcre update-pcre tests
+SUBDIRS = libcharset $(PRINTF_SUBDIR) $(MAYBE_PCRE) update-pcre .
+DIST_SUBDIRS = libcharset gnulib pcre update-pcre
 
 AM_CPPFLAGS = 				\
 	$(glib_INCLUDES) 		\
--- a/Makefile.am
+++ b/Makefile.am
@@ -6,8 +6,8 @@
 
 ACLOCAL_AMFLAGS = -I m4macros ${ACLOCAL_FLAGS}
 
-SUBDIRS = . m4macros glib gmodule gthread gobject gio po docs tests
-DIST_SUBDIRS = $(SUBDIRS) build win32
+SUBDIRS = . m4macros glib gmodule gthread gobject gio
+DIST_SUBDIRS = $(SUBDIRS) build
 
 bin_SCRIPTS = glib-gettextize
 
--- a/configure.ac
+++ b/configure.ac
@@ -63,7 +63,7 @@
 # Save this value here, since automake will set cflags later
 cflags_set=${CFLAGS:+set}
 
-AM_INIT_AUTOMAKE([1.13.3 -Wno-portability no-define no-dist-gzip dist-xz tar-ustar subdir-objects])
+AM_INIT_AUTOMAKE([1.15 -Wno-portability no-define no-dist-gzip dist-xz tar-ustar subdir-objects])
 AM_MAINTAINER_MODE([enable])
 
 # Support silent build rules. Disable

