--- tremor-20040525/vorbisidec.pc	1970-01-01 01:00:00.000000000 +0100
+++ tremor-20040525.new/vorbisidec.pc	2003-09-10 17:51:26.000000000 +0200
@@ -0,0 +1,11 @@
+prefix=
+exec_prefix=${prefix}
+libdir=${exec_prefix}/lib
+includedir=${prefix}/include
+
+Name: vorbisidec
+Description: OGG Vorbis Audio Decoder
+Requires:
+Version: 20040525
+Libs: -L${libdir} -lvorbisidec
+Cflags: -I${includedir}
--- libvorbisidec-1.0.2+svn14261.org/configure.in	2007-07-24 00:35:47.000000000 +0200
+++ libvorbisidec-1.0.2+svn14261/configure.in	2009-03-28 20:14:54.000000000 +0100
@@ -5,6 +5,7 @@
 dnl ------------------------------------------------
 
 AC_INIT(mdct.c)
+AC_CONFIG_AUX_DIR(.)
 
 AC_CANONICAL_HOST
 AC_CANONICAL_TARGET
--- libvorbisidec-1.0.2+svn14261.org/configure.in	2007-07-24 00:35:47.000000000 +0200
+++ libvorbisidec-1.0.2+svn14261/configure.in	2009-03-28 20:14:54.000000000 +0100
@@ -10,7 +10,7 @@
 AC_CANONICAL_HOST
 AC_CANONICAL_TARGET

-AM_CONFIG_HEADER([config.h])
+AC_CONFIG_HEADERS([config.h])

 AM_INIT_AUTOMAKE(libvorbisidec,1.2.1)
