--- a/libxml.h
+++ b/libxml.h
@@ -13,10 +13,13 @@
 
 #ifndef NO_LARGEFILE_SOURCE
 #ifndef _LARGEFILE_SOURCE
 #define _LARGEFILE_SOURCE
 #endif
+#ifndef _LARGEFILE64_SOURCE
+#define _LARGEFILE64_SOURCE
+#endif
 #ifndef _FILE_OFFSET_BITS
 #define _FILE_OFFSET_BITS 64
 #endif
 #endif
 
--- a/Makefile.am
+++ b/Makefile.am
@@ -2,9 +2,9 @@
 
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = include . doc example xstc $(PYTHON_SUBDIR)
-
-DIST_SUBDIRS = include . doc example python xstc
+SUBDIRS = include . $(PYTHON_SUBDIR)
+
+DIST_SUBDIRS = include . python
 
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(srcdir)/include
 
