--- libxml2-2.9.0.org/configure.in
+++ libxml2-2.9.0/configure.in
@@ -812,7 +812,7 @@ if test "$with_python" != "no" ; then
 		    PYTHON_SITE_PACKAGES=`$PYTHON -c "from distutils import sysconfig; print sysconfig.get_python_lib()"`
 	    fi
 	fi
-	PYTHON_LIBS=`python$PYTHON_VERSION-config --ldflags`
+	PYTHON_LIBS=''
     fi
     if test "$with_python" != ""
     then
--- libxml2-2.9.0.orig/libxml.h
+++ libxml2-2..0/libxml.h
@@ -11,10 +11,13 @@
 
 #ifndef NO_LARGEFILE_SOURCE
 #ifndef _LARGEFILE_SOURCE
 #define _LARGEFILE_SOURCE
 #endif
+#ifndef _LARGEFILE64_SOURCE
+#define _LARGEFILE64_SOURCE
+#endif
 #ifndef _FILE_OFFSET_BITS
 #define _FILE_OFFSET_BITS 64
 #endif
 #endif
 
--- libxml2-2.9.0/Makefile.am.org
+++ libxml2-2.9.0/Makefile.am
@@ -2,9 +2,9 @@
 
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = include . doc example xstc $(PYTHON_SUBDIR)
-
-DIST_SUBDIRS = include . doc example python xstc
+SUBDIRS = include . $(PYTHON_SUBDIR)
+
+DIST_SUBDIRS = include . python
 
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(srcdir)/include
 
