--- a/configure.ac
+++ b/configure.ac
@@ -44,7 +44,7 @@
 AC_SUBST(LIBXML_VERSION_NUMBER)
 AC_SUBST(LIBXML_VERSION_EXTRA)
 
-AM_INIT_AUTOMAKE([1.16.3 foreign subdir-objects no-dist-gzip dist-xz])
+AM_INIT_AUTOMAKE([1.15.1 foreign subdir-objects no-dist-gzip dist-xz])
 AM_MAINTAINER_MODE([enable])
 AM_SILENT_RULES([yes])
 
--- a/Makefile.am
+++ b/Makefile.am
@@ -2,7 +2,7 @@
 
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = include . example
+SUBDIRS = include .
 if WITH_DOXYGEN
 SUBDIRS += doc
 endif
@@ -13,7 +13,7 @@
 SUBDIRS += fuzz
 endif
 
-DIST_SUBDIRS = include . doc example fuzz python
+DIST_SUBDIRS = include .
 
 AM_CPPFLAGS = -I$(builddir)/include -I$(srcdir)/include
 
