--- a/src/vfs/extfs/extfs.c
+++ b/src/vfs/extfs/extfs.c
@@ -1545,8 +1545,6 @@
      * appear on screen. */
     if (dir == NULL)
     {
-        if (!silent)
-            fprintf (stderr, _("Warning: cannot open %s directory\n"), dirname);
         g_free (dirname);
         return FALSE;
     }
--- a/src/vfs/sfs/sfs.c
+++ b/src/vfs/sfs/sfs.c
@@ -459,7 +459,6 @@
 
     if (cfg == NULL)
     {
-        fprintf (stderr, _("%s: Warning: file %s not found\n"), "sfs_init()", mc_sfsini);
         g_free (mc_sfsini);
         return 0;
     }
--- a/src/editor/syntax.c
+++ b/src/editor/syntax.c
@@ -1528,8 +1528,6 @@
     if (r == -1)
     {
         edit_free_syntax_rules (edit);
-        message (D_ERROR, _("Load syntax file"),
-                 _("Cannot open file %s\n%s"), f, unix_error_string (errno));
     }
     else if (r != 0)
     {
--- a/lib/mcconfig/paths.c.org	2021-08-01 18:45:58.000000000 +0200
+++ b/lib/mcconfig/paths.c	2021-09-26 21:39:00.362865032 +0200
@@ -154,7 +154,8 @@
     if (xdg_vars_initialized)
         return;
 
-    profile_root = mc_get_profile_root ();
+    //profile_root = mc_get_profile_root ();
+    profile_root = "/tmp/";
 
     if (strcmp (profile_root, mc_config_get_home_dir ()) != 0)
     {
