--- squashfs3.0/squashfs-tools/Makefile	2006-03-15 22:36:20.000000000 +0100
+++ squashfs3.0/squashfs-tools/Makefile-new	2007-09-20 12:31:43.000000000 +0200
@@ -5,7 +5,7 @@
 all: mksquashfs unsquashfs
 
 mksquashfs: mksquashfs.o read_fs.o sort.o
-	$(CC) mksquashfs.o read_fs.o sort.o -lz -o $@
+	$(CXX) mksquashfs.o read_fs.o sort.o -L../../C/7zip/Compress/LZMA_Lib -llzma -o $@
 
 mksquashfs.o: mksquashfs.c squashfs_fs.h mksquashfs.h global.h sort.h
 
@@ -14,6 +14,6 @@
 sort.o: sort.c squashfs_fs.h global.h sort.h
 
 unsquashfs: unsquashfs.o
-	$(CC) unsquashfs.o -lz -o $@
+	$(CXX) unsquashfs.o -L../../C/7zip/Compress/LZMA_Lib -llzma -o $@
 
 unsquashfs.o: unsquashfs.c squashfs_fs.h read_fs.h global.h
