--- a/Makefile.in.org
+++ b/Makefile.in
@@ -135,10 +135,6 @@
 install: progs $(srcdir)/doc/socat.1
 	mkdir -p $(DESTDIR)$(BINDEST)
 	$(INSTALL) -m 755 socat $(DESTDIR)$(BINDEST)
-	$(INSTALL) -m 755 procan $(DESTDIR)$(BINDEST)
-	$(INSTALL) -m 755 filan $(DESTDIR)$(BINDEST)
-	mkdir -p $(DESTDIR)$(MANDEST)/man1
-	$(INSTALL) -m 644 $(srcdir)/doc/socat.1 $(DESTDIR)$(MANDEST)/man1/
 
 uninstall:
 	rm -f $(DESTDIR)$(BINDEST)/socat
