--- a/tuxtxt.c
+++ b/tuxtxt.c
@@ -225,13 +225,21 @@
 		}
 		if (ioctl(rc[rc_num], EVIOCGNAME(128), tmp) < 0)
 			perror("EVIOCGNAME failed");
+#ifdef __sh__
+		if (!strstr(tmp, "TDT RC event driver") && !strstr(tmp, "key"))
+#else
 		if (!strstr(tmp, "remote control") && !strstr(tmp, "key"))
+#endif
 		{
 			close(rc[rc_num]);
 			rc[rc_num] = -1;
 		}
 		else
+#ifdef __sh__
+			break;
+#else
 			++rc_num;
+#endif
 		++cnt;
 	}
 
@@ -2943,6 +2951,9 @@
 				case KEY_MENU:		RCCode = RC_DBOX;	break;
 				case KEY_TEXT:
 				case KEY_EXIT:		RCCode = RC_HOME;	break;
+#ifdef __sh__
+				case KEY_HOME:		RCCode = RC_HOME;	break;
+#endif
 				case KEY_POWER:		RCCode = RC_STANDBY;	break;
 				default:			RCCode = -1;		break;
 				}
