--- a/Makefile
+++ b/Makefile
@@ -101,12 +101,15 @@
   WEDEF_FLAG= -DWE_ESSENTIAL=y
 endif
 
+ifndef OPT_FLAGS
+  OPT_FLAGS=
+endif
 # Other flags
 CFLAGS=-Os -W -Wall -Wstrict-prototypes -Wmissing-prototypes -Wshadow \
 	-Wpointer-arith -Wcast-qual -Winline -I.
 #CFLAGS=-O2 -W -Wall -Wstrict-prototypes -I.
 DEPFLAGS=-MMD
-XCFLAGS=$(CFLAGS) $(DEPFLAGS) $(WARN) $(HEADERS) $(WELIB_FLAG) $(WEDEF_FLAG)
+XCFLAGS=$(CFLAGS) $(DEPFLAGS) $(WARN) $(HEADERS) $(WELIB_FLAG) $(WEDEF_FLAG) $(OPT_FLAGS)
 PICFLAG=-fPIC
 
 # Standard compilation targets
@@ -146,6 +149,7 @@
 # Compilation of the dynamic library
 $(DYNAMIC): $(OBJS:.o=.so)
 	$(CC) -shared -o $@ -Wl,-soname,$@ $(STRIPFLAGS) $(LIBS) -lc $^
+	ln -s $@ $(DYNAMIC_LINK)
 
 # Compilation of the static library
 $(STATIC): $(OBJS:.o=.so)
--- a/iwlib.c
+++ b/iwlib.c
@@ -667,6 +667,7 @@
 {
   struct iwreq		wrq;
 
+  memset((char *) &wrq, 0, sizeof(struct iwreq));
   memset((char *) info, 0, sizeof(struct wireless_config));
 
   /* Get wireless name */
