--- a/arch/sh/kernel/process_32.c
+++ b/arch/sh/kernel/process_32.c
@@ -37,21 +37,16 @@
 
 int ubc_usercnt = 0;
 
-static void watchdog_trigger_immediate(void)
-{
-	sh_wdt_write_cnt(0xFF);
-	sh_wdt_write_csr(0xC2);
-}
-
 void machine_restart(char * __unused)
 {
-	local_irq_disable();
-
-	/* Use watchdog timer to trigger reset */
-	watchdog_trigger_immediate();
-
-	while (1)
-		cpu_sleep();
+	/* Taken from stlinux22 because stlinux24
+	 * handling does not work on ufs910.
+	 */
+	
+	/* SR.BL=1 and invoke address error to let CPU reset (manual reset) */
+	asm volatile ("ldc %0, sr\n\t"
+		      "mov.l @%1, %0"::"r" (0x10000000), "r"(0x80000001));
+ 
 }
 
 void machine_halt(void)
