--- a/drivers/media/dvb/dvb-core/dvb_ca_en50221.c
+++ b/drivers/media/dvb/dvb-core/dvb_ca_en50221.c
@@ -380,7 +380,8 @@
  * @return 0 on success, nonzero on error.
  */
 static int dvb_ca_en50221_read_tuple(struct dvb_ca_private *ca, int slot,
-				     int *address, int *tupleType, int *tupleLength, u8 * tuple)
+				     int *address, int *tupleType, int *tupleLength, u8 * tuple,
+				     unsigned int tupleTypeDelay, unsigned int tupleLengthDelay)
 {
 	int i;
 	int _tupleType;
@@ -397,11 +398,16 @@
 		*tupleLength = 0;
 		return 0;
 	}
+	if(tupleTypeDelay)
+		msleep(tupleTypeDelay);
+
 	if ((_tupleLength = ca->pub->read_attribute_mem(ca->pub, slot, _address + 2)) < 0)
 		return _tupleLength;
 	_address += 4;
 
 	dprintk("TUPLE type:0x%x length:%i\n", _tupleType, _tupleLength);
+	if(tupleLengthDelay)
+		msleep(tupleLengthDelay);
 
 	/* read in the whole tuple */
 	for (i = 0; i < _tupleLength; i++) {
@@ -447,7 +453,7 @@
 
 	// CISTPL_DEVICE_0A
 	if ((status =
-	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple)) < 0)
+	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple, 500, 10)) < 0)
 		return status;
 	if (tupleType != 0x1D)
 		return -EINVAL;
@@ -456,7 +462,7 @@
 
 	// CISTPL_DEVICE_0C
 	if ((status =
-	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple)) < 0)
+	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple, 0, 0)) < 0)
 		return status;
 	if (tupleType != 0x1C)
 		return -EINVAL;
@@ -465,7 +471,7 @@
 
 	// CISTPL_VERS_1
 	if ((status =
-	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple)) < 0)
+	     dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType, &tupleLength, tuple, 0, 0)) < 0)
 		return status;
 	if (tupleType != 0x15)
 		return -EINVAL;
@@ -474,7 +480,7 @@
 
 	// CISTPL_MANFID
 	if ((status = dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType,
-						&tupleLength, tuple)) < 0)
+						&tupleLength, tuple, 0, 0)) < 0)
 		return status;
 	if (tupleType != 0x20)
 		return -EINVAL;
@@ -487,7 +493,7 @@
 
 	// CISTPL_CONFIG
 	if ((status = dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType,
-						&tupleLength, tuple)) < 0)
+						&tupleLength, tuple, 0, 0)) < 0)
 		return status;
 	if (tupleType != 0x1A)
 		return -EINVAL;
@@ -520,7 +526,7 @@
 	/* process the CFTABLE_ENTRY tuples, and any after those */
 	while ((!end_chain) && (address < 0x1000)) {
 		if ((status = dvb_ca_en50221_read_tuple(ca, slot, &address, &tupleType,
-							&tupleLength, tuple)) < 0)
+							&tupleLength, tuple, 0, 0)) < 0)
 			return status;
 		switch (tupleType) {
 		case 0x1B:	// CISTPL_CFTABLE_ENTRY
