diff --git a/packages/zlib/package.desc b/packages/zlib/package.desc
index d85e5656..1aaa84a6 100644
--- a/packages/zlib/package.desc
+++ b/packages/zlib/package.desc
@@ -1,4 +1,4 @@
 repository='git https://github.com/madler/zlib.git'
-mirrors='http://downloads.sourceforge.net/project/libpng/zlib/${CT_ZLIB_VERSION} https://www.zlib.net/'
+mirrors='http://downloads.sourceforge.net/project/libpng/zlib/${CT_ZLIB_VERSION} https://www.zlib.net/fossils/'
 archive_formats='.tar.xz .tar.gz'
 signature_format='packed/.asc'
