*** linux-sh4-2.6.23.17_stm23_0123_orig/drivers/mtd/maps/physmap.c	2011-06-19 18:34:29.000000000 +0300
--- linux-sh4/drivers/mtd/maps/physmap.c	2011-07-29 06:50:51.114192548 +0300
@@ -79,10 +79,97 @@
 					"qinfo_probe",
 					"map_rom",
 					NULL };
-#ifdef CONFIG_MTD_PARTITIONS
-static const char *part_probe_types[] = { "cmdlinepart", "RedBoot", NULL };
-#endif
+
+static struct mtd_partition partition_info[]=
+{
+	{
+	 .name   = "Boot_firmware", //mtd0
+	 .size   = 0x00060000,  //384k
+	 .offset = 0x00000000,
+	 },
+	{
+	 .name   = "kernel",    //mtd1
+	 .size   = 0x00320000, //3,125M
+	 .offset = 0x00060000, //384k (0.375M)
+	 },
+	{
+	 .name   = "ROOT_FS",   //mtd2
+	 .size   = 0x00020000,  //128k (dummy)
+	 .offset = 0x00380000,  //3.5M
+	 },
+	{
+	 .name   = "Device",    //mtd3
+	 .size   = 0x00020000,  //128k (dummy)
+	 .offset = 0x003A0000,  //3.625M
+	 },
+	{
+	 .name   = "APP",       //mtd4
+	 .size   = 0x00020000,  //128k (dummy)
+	 .offset = 0x003C0000,  //3.75M
+	 },
+	{
+	 .name   = "Real_ROOT", //mtd5
+	 .size   = 0x01D00000,  //29Mbyte, 128k hole at 0x20FE0000 is used for force flash
+	 .offset = 0x003E0000,  //3.875M
+	 },
+	{
+	 .name   = "Config",    //mtd6
+	 .size   = 0x00100000,  //1M
+	 .offset = 0x02100000,
+	 },
+	{
+	 .name   = "User",      //mtd7
+	 .size   = 0x01E00000,  //30M
+	 .offset = 0x02200000,
+	 },
+};
+
+static struct mtd_partition higlvl_partition_info[]=
+{
+	{
+	 .name   = "Boot_firmware", //mtd0
+	 .size   = 0x00060000,	//384k (0.375M)
+	 .offset = 0x00000000,
+	 },
+	{
+	 .name   = "Kernel",    //mtd1
+	 .size   = 0x00320000,  //3,125M
+	 .offset = 0x00060000,  //0.375M
+	 },
+	{
+	 .name   = "Fake_ROOT", //mdt2
+	 .size   = 0x0001FFFE,  //128k (dummy) minus 1 word to force read only mount
+	 .offset = 0x00380000,  //4M - 128k force flash hole - 128k Fake_APP - 128k Fake_DEV - 128k own size
+	 },
+	{
+	 .name   = "Fake_DEV",  //mtd3
+	 .size   = 0x0001FFFE,  //128k (dummy) minus 1 word to force read only mount
+	 .offset = 0x003A0000,  //4M - 128k force flash hole - 128k Fake_APP - 128k own size
+	 },
+	{
+	 .name   = "Fake_APP",  //mtd4
+	 .size   = 0x0001FFFE,  //128k (dummy) minus 1 word to force read only mount
+	 .offset = 0x003C0000,  //4M - 128k force flash hole - 128k own size
+	 },
+	{
+	 .name   = "Real_ROOT", //mtd5
+	 .size   = 0x03C00000,  //60M
+	 .offset = 0x003E0000,  //4M - 128k force flash hole
+	},
+//	{
+//	 .name = "SWAP",        //mtd6
+//	 .size   = 0x003E0000,  //4M - one erase block, kernel complains at 4M exactly
+//	 .offset = 0x03C00000,  //60M
+//	 },
+};
+
+#define NUM_PARTITIONS ARRAY_SIZE(partition_info)
+#define NUM_HIGHLVL_PARTITIONS ARRAY_SIZE(higlvl_partition_info)
+
+static struct mtd_info *lowlvl_parts[NUM_PARTITIONS];
+static struct mtd_info *merged_mtd;
+
 
 static int physmap_flash_probe(struct platform_device *dev)
 {
 	struct physmap_flash_data *physmap_data;
@@ -106,7 +195,7 @@
 	platform_set_drvdata(dev, info);
 
 	for (i = 0; i < dev->num_resources; i++) {
-		printk(KERN_NOTICE "physmap platform flash device: %.8llx at %.8llx\n",
+		printk("physmap platform flash device: %.8llx at %.8llx\n",
 		       (unsigned long long)(dev->resource[i].end - dev->resource[i].start + 1),
 		       (unsigned long long)dev->resource[i].start);
 
@@ -126,6 +215,8 @@
 		info->map[i].set_vpp = physmap_data->set_vpp;
 		info->map[i].pfow_base = physmap_data->pfow_base;
 
+		printk("MAP[%d]: %s %p %lu %d %p\n", i, info->map[i].name, (void *)info->map[i].phys, info->map[i].size, info->map[i].bankwidth, info->map[i].set_vpp);
+
 		info->map[i].virt = devm_ioremap(&dev->dev, info->map[i].phys,
 						 info->map[i].size);
 		if (info->map[i].virt == NULL) {
@@ -141,6 +230,8 @@
 #endif
 		simple_map_init(&info->map[i]);
 
+		printk("MAP[%d]: virtual address: 0x%p\n", i, info->map[i].virt);
+
 		probe_type = rom_probe_types;
 		for (; info->mtd[i] == NULL && *probe_type != NULL; probe_type++)
 			info->mtd[i] = do_map_probe(*probe_type, &info->map[i]);
@@ -174,6 +265,25 @@
 	if (err)
 		goto err_out;
 
+	partition_info[0].mtdp = &lowlvl_parts[0];
+	partition_info[1].mtdp = &lowlvl_parts[1];
+	partition_info[2].mtdp = &lowlvl_parts[2];
+	partition_info[3].mtdp = &lowlvl_parts[3];
+	partition_info[4].mtdp = &lowlvl_parts[4];
+	partition_info[5].mtdp = &lowlvl_parts[5];
+	partition_info[6].mtdp = &lowlvl_parts[6];
+	partition_info[7].mtdp = &lowlvl_parts[7];
+
+	add_mtd_partitions(info->cmtd, partition_info, NUM_PARTITIONS);
+
+	merged_mtd = mtd_concat_create(lowlvl_parts, NUM_PARTITIONS, "NOR Flash Concatenated");
+	if(merged_mtd)
+	{
+		add_mtd_partitions(merged_mtd, higlvl_partition_info, NUM_HIGHLVL_PARTITIONS);
+	}
+
+
+#if 0
 #ifdef CONFIG_MTD_PARTITIONS
 	err = parse_mtd_partitions(info->cmtd, part_probe_types,
 				&info->parts, 0);
@@ -192,6 +302,7 @@
 #endif
 
 	add_mtd_device(info->cmtd);
+#endif
 	return 0;
 
 err_out:
*** linux-sh4-2.6.23.17_stm23_0123_orig/drivers/mtdpart.c	2011-06-19 18:34:29.000000000 +0300
--- linux-sh4/drivers/mtd/mtdpart.c	2011-07-29 06:50:51.114192548 +0300
@@ -27,7 +27,9 @@
 	struct mtd_info mtd;
 	struct mtd_info *master;
 	uint64_t offset;
+	int index;
 	struct list_head list;
+	int registered;
 };
 
 /*
@@ -345,7 +347,8 @@
 	list_for_each_entry_safe(slave, next, &mtd_partitions, list)
 		if (slave->master == master) {
 			list_del(&slave->list);
+			if (slave->registered)
+				del_mtd_device(&slave->mtd);
-			del_mtd_device(&slave->mtd);
 			kfree(slave);
 		}
 
@@ -438,6 +441,7 @@
 	slave->mtd.erase = part_erase;
 	slave->master = master;
 	slave->offset = part->offset;
+	slave->index = partno;
 
 	if (slave->offset == MTDPART_OFS_APPEND)
 		slave->offset = cur_offset;
@@ -530,9 +534,15 @@
 		slave->mtd.flags |= MTD_SPANS_MASTER;
 
 out_register:
+	if (part->mtdp) {
+		/* store the object pointer (caller may or may not register it*/
+		*part->mtdp = &slave->mtd;
+		slave->registered = 0;
+	} else {
+		/* register our partition */
+		add_mtd_device(&slave->mtd);
+		slave->registered = 1;
+	}
-	/* register our partition */
-	add_mtd_device(&slave->mtd);
-
 	return slave;
 }
 
*** linux-sh4-2.6.23.17_stm23_0123_orig/include/linux/mtd/partitions.h	2011-06-19 18:34:29.000000000 +0300
--- linux-sh4/include/linux/mtd/partitions.h	2011-07-29 06:50:51.114192548 +0300
@@ -40,6 +40,7 @@
 	uint64_t offset;		/* offset within the master MTD space */
 	uint32_t mask_flags;		/* master MTD flags to mask out for this partition */
 	struct nand_ecclayout *ecclayout;	/* out of band layout for this partition (NAND only)*/
+	struct mtd_info **mtdp;		/* pointer to store the MTD object */
 };
 
 #define MTDPART_OFS_NXTBLK	(-2)
