--- a/arch/sh/boards/mach-mb442/setup.c	2011-01-20 19:00:37.000000000 +0100
+++ b/arch/sh/boards/mach-mb442/setup.c	2011-02-01 22:47:01.403481000 +0100
@@ -21,7 +21,9 @@
 #include <linux/mtd/physmap.h>
 #include <linux/stm/platform.h>
 #include <linux/stm/stx7100.h>
+#include <linux/bpa2.h>
 #include <asm/irl.h>
+#include <linux/smsc911x.h>
 
 
 
@@ -29,7 +31,25 @@
 #define MB442_PIO_SMC91X_RESET stm_gpio(2, 6)
 #define MB442_PIO_FLASH_VPP stm_gpio(2, 7)
 
+const char *LMI_VID_partalias[] = { "BPA2_Region1", "coredisplay-video", "gfx-memory", "v4l2-video-buffers", "v4l2-coded-video-buffers", NULL };
+const char *LMI_SYS_partalias[] = { "BPA2_Region0", "bigphysarea", NULL };
 
+static struct bpa2_partition_desc bpa2_parts_table[] = {
+      {
+	      .name  = "LMI_VID",
+	      .start = 0x10800000,
+	      .size  = 0x03800000,
+	      .flags = 0,
+	      .aka   = LMI_VID_partalias
+      },
+      {
+	      .name  = "LMI_SYS",
+	      .start = 0,
+	      .size  = 0x00E00000,
+	      .flags = 0,
+	      .aka   = LMI_SYS_partalias
+      }
+};
 
 void __init mb442_setup(char** cmdline_p)
 {
@@ -43,26 +63,34 @@
 	stx7100_configure_asc(3, &(struct stx7100_asc_config) {
 			.hw_flow_control = 0,
 			.is_console = 0, });
+
+  	bpa2_init(bpa2_parts_table, ARRAY_SIZE(bpa2_parts_table));
 }
 
-static struct resource mb442_smc91x_resources[] = {
+static struct resource smsc_lan9115_resources[] = {
 	[0] = {
-		.start	= 0x02000300,
-		.end	= 0x02000300 + 0xff,
-		.flags	= IORESOURCE_MEM,
-	},
+ 		      .flags = IORESOURCE_MEM,
+ 		      .start = 0x02000000,
+ 		      .end   = 0x020000ff,
+ 	      },
 	[1] = {
-		.start	= IRL0_IRQ,
-		.end	= IRL0_IRQ,
-		.flags	= IORESOURCE_IRQ,
-	},
+ 		      .flags = IORESOURCE_IRQ,
+ 		      .start = IRL0_IRQ,
+ 		      .end   = IRL0_IRQ,
+ 	      },
 };
 
-static struct platform_device mb442_smc91x_device = {
-	.name		= "smc91x",
+static struct platform_device smsc_lan9115 = {
+	.name		= "smsc911x",
 	.id		= 0,
-	.num_resources	= ARRAY_SIZE(mb442_smc91x_resources),
-	.resource	= mb442_smc91x_resources,
+	.num_resources	= ARRAY_SIZE(smsc_lan9115_resources),
+	.resource	= smsc_lan9115_resources,
+	.dev.platform_data = &(struct smsc911x_platform_config) {
+		.irq_polarity = 1,
+		.irq_type = 1,
+		.flags = SMSC911X_USE_32BIT,
+		.phy_interface = PHY_INTERFACE_MODE_MII,
+	},
 };
 
 static void mb442_set_vpp(struct map_info *info, int enable)
@@ -70,28 +98,66 @@
 	gpio_set_value(MB442_PIO_FLASH_VPP, enable);
 }
 
+static struct resource physmap_flash_resource = {
+	.start		= 0x00000000,
+	.end		= 0x01000000 - 1,
+	.flags		= IORESOURCE_MEM,
+};
+
+static struct mtd_partition mtd_parts_table[7] = {
+    {
+	.name = "Boot firmware",
+	.size = 0x00020000,
+	.offset = 0x00000000
+    },
+    {
+	.name = "Kernel - RAW",
+	.size = 0x00190000,
+	.offset = 0x00040000
+    },
+    {
+	.name = "ROOT - SQUASHFS",
+	.size = 0x00B40000,
+	.offset = 0x001D0000
+    },
+    {
+	.name = "Var - JFFS2",
+	.size = 0x002F0000,
+	.offset = 0x00D10000
+    },
+    {
+	.name = "BootConfiguration",
+	.size = 0x00010000,
+	.offset = 0x00020000
+    },
+    {
+	.name = "Full without bootloader",
+	.size = 0x00fc0000,
+	.offset = 0x00040000
+    },
+    {
+	.name = "MiniUboot - RAW",
+	.size = 0x00010000,
+	.offset = 0x00030000
+    }
+};
+
 static struct platform_device mb442_physmap_flash = {
 	.name		= "physmap-flash",
 	.id		= -1,
 	.num_resources	= 1,
-	.resource	= (struct resource[]) {
-		STM_PLAT_RESOURCE_MEM(0, 8*1024*1024),
-	},
+	.resource	= &physmap_flash_resource,
 	.dev.platform_data = &(struct physmap_flash_data) {
 		.width		= 2,
 		.set_vpp	= mb442_set_vpp,
+		.nr_parts       = ARRAY_SIZE(mtd_parts_table),
+		.parts		= mtd_parts_table,
 	},
 };
 
 
 static int mb442_phy_reset(void* bus)
 {
-	gpio_set_value(MB442_PIO_STE100P_RESET, 1);
-	udelay(1);
-	gpio_set_value(MB442_PIO_STE100P_RESET, 0);
-	udelay(1);
-	gpio_set_value(MB442_PIO_STE100P_RESET, 1);
-
 	return 1;
 }
 
@@ -106,9 +172,15 @@
 	.irqs = stmmac_phy_irqs,
 };
 
+static struct platform_device led_device = {
+       .name		= "ufs910-led",
+       .id		= -1,
+};
+
 static struct platform_device *mb442_devices[] __initdata = {
-	&mb442_smc91x_device,
+	&smsc_lan9115,
 	&mb442_physmap_flash,
+	&led_device,
 };
 
 static int __init mb442_device_init(void)
@@ -120,7 +192,7 @@
 			.out1_enabled = 1, });
 
 	stx7100_configure_ssc_i2c(0, NULL);
-	stx7100_configure_ssc_spi(1, NULL);
+	stx7100_configure_ssc_i2c(1, NULL);
 	stx7100_configure_ssc_i2c(2, NULL);
 
 	stx7100_configure_usb();
@@ -138,25 +210,6 @@
 	gpio_request(MB442_PIO_FLASH_VPP, "Flash VPP");
 	gpio_direction_output(MB442_PIO_FLASH_VPP, 0);
 
-	gpio_request(MB442_PIO_STE100P_RESET, "STE100P reset");
-	gpio_direction_output(MB442_PIO_STE100P_RESET, 1);
-
-	stx7100_configure_ethernet(&(struct stx7100_ethernet_config) {
-			.mode = stx7100_ethernet_mode_mii,
-			.ext_clk = 0,
-			.phy_bus = 0,
-			.phy_addr = STMMAC_PHY_ADDR,
-			.mdio_bus_data = &stmmac_mdio_bus,
-		});
-
-	/* Reset the SMSC 91C111 Ethernet chip */
-	gpio_request(MB442_PIO_SMC91X_RESET, "SMC91x reset");
-	gpio_direction_output(MB442_PIO_SMC91X_RESET, 0);
-	udelay(1);
-	gpio_set_value(MB442_PIO_SMC91X_RESET, 1);
-	udelay(1);
-	gpio_set_value(MB442_PIO_SMC91X_RESET, 0);
-
 	return platform_add_devices(mb442_devices,
 			ARRAY_SIZE(mb442_devices));
 }
--- a/drivers/stm/stx7100_comms.c	2011-01-20 19:00:37.000000000 +0100
+++ b/drivers/stm/stx7100_comms.c	2011-02-01 22:52:17.018479519 +0100
@@ -226,12 +226,10 @@
 			STM_PAD_PIO_BIDIR_NAMED(2, 0, 1, "SCL"),
 			STM_PAD_PIO_BIDIR_NAMED(2, 1, 1, "SDA"),
 		},
-		.sysconfs_num = 2,
+		.sysconfs_num = 1,
 		.sysconfs = (struct stm_pad_sysconf []) {
 			/* SSC0_MUX_SEL = 0 (default assignment) */
 			STM_PAD_SYS_CFG(7, 1, 1, 0),
-			/* DVO_OUT_ON = 0 (SSC not DVO) */
-			STM_PAD_SYS_CFG(7, 10, 10, 0),
 		},
 	},
 	[1] = {
@@ -345,7 +343,7 @@
 	BUG_ON(stx7100_ssc_configured[ssc]);
 	stx7100_ssc_configured[ssc] = 1;
 
-	stx7100_ssc_devices[ssc].name = "i2c-stm";
+	stx7100_ssc_devices[ssc].name = "i2c-st40";
 	stx7100_ssc_devices[ssc].id = i2c_busnum;
 
 	plat_data = stx7100_ssc_devices[ssc].dev.platform_data;
